Core/include/Acts/EventData/MultiTrajectory.hpp

```diff
@@ -328,23 +271,87 @@ class TrackStateProxy {
     setReferenceSurface(other.referenceSurfacePointer());
   }
 
-  /// Return the index tuple that makes up this track state
-  /// @return Immutable ref to index tuple from the parent @c MultiTrajectory
-  const IndexData& data() const { return m_traj->m_index[m_istate]; }
+  /// Unset an optional track state component
+  /// @param target The component to unset
+  template <bool RO = ReadOnly, typename = std::enable_if<!RO>>
+  void unset(TrackStatePropMask target) {
+    m_traj->unset(target, m_istate);
+  }
 
   /// Reference surface.
   /// @return the reference surface
   const Surface& referenceSurface() const {
-    assert(data().irefsurface != IndexData::kInvalid);
-    return *m_traj->m_referenceSurfaces[data().irefsurface];
+    assert(has<hashString("referenceSurface")>());
+    return *component<std::shared_ptr<const Surface>,
+                      hashString("referenceSurface")>();
   }
 
   /// Set the reference surface to a given value
   /// @param srf Shared pointer to the surface to set
   /// @note This overload is only present in case @c ReadOnly is false.
   template <bool RO = ReadOnly, typename = std::enable_if_t<!RO>>
   void setReferenceSurface(std::shared_ptr<const Surface> srf) {
-    m_traj->m_referenceSurfaces[data().irefsurface] = std::move(srf);
+    assert(has<hashString("referenceSurface")>());
```

Comment:
Not sure. In many other places we also only have checks via `assert` i.e. only `#ifndef NDEBUG`. 
In this particular case I think the check is obsolete now anyway because it `referenceSurface` will always be present.

Reply:
I removed this `has` check, `referenceSurface` should always be available for setting.